framework/src/org/checkerframework/framework/util/expressionannotations/ExpressionAnnotationHelper.java

```diff
@@ -0,0 +1,715 @@
+        }
+
+        private List<ExpressionAnnotationError> checkForError(AnnotationMirror am) {
+            List<String> value =
+                    AnnotationUtils.getElementValueArray(am, "value", String.class, true);
+            List<ExpressionAnnotationError> errors = new ArrayList<>();
+            for (String v : value) {
+                if (ExpressionAnnotationError.isExpressionError(v)) {
+                    errors.add(new ExpressionAnnotationError(v));
+                }
+            }
+            return errors;
+        }
+    }
+
+    /** Copies annotations that might have been viewpoint adapted from type to the parameter. */
+    private class ViewpointAdaptedCopier extends AnnotatedTypeComparer<Void> {
+        @Override
+        protected Void scan(AnnotatedTypeMirror type, AnnotatedTypeMirror p) {
+            if (type == null || p == null) {
+                return null;
+            }
+            Set<AnnotationMirror> replacement = AnnotationUtils.createAnnotationSet();
+            for (Class<? extends Annotation> vpa : expressionAnnos) {
+                AnnotationMirror anno = type.getAnnotation(vpa);
+                if (anno != null) {
+                    // Only replace annotations that might have been changed.
+                    replacement.add(anno);
+                }
+            }
+            p.replaceAnnotations(replacement);
+            if (type.getKind() != p.getKind()) {
+                // if the underlying types don't match, then this type has be substituted for a
+                // type variable, so don't recur. The primary annotation was copied because
+                // if the type variable might have had a primary annotation at a use.
+                // For example:
+                // <T> void method(@KeyFor("a") T t) {...}
+                // void use(@KeyFor("b") String s) {
+                //      method(s);  // the type of the parameter should be @KeyFor("a") String
+                // }
+                return null;
+            }
+            return super.scan(type, p);
+        }
+
+        @Override
+        protected Void compare(AnnotatedTypeMirror type, AnnotatedTypeMirror p) {
+            if (type == null || p == null) {
+                return null;
+            }
+            if (type.getKind() != p.getKind()) {
+                ErrorReporter.errorAbort("Should be the same. type: %s p: %s ", type, p);
+            }
+            return null;
+        }
+
+        @Override
+        protected Void combineRs(Void r1, Void r2) {
+            return null;
+        }
+    }
+
+    /**
+     * Whether or not atm has an expression annotation. If an annotated type does not have an
+     * expression annotation, then no standardization or viewpoint adaption is performed. (This
+     * check avoids calling time intensive methods unless absolutely required.)
+     */
+    private boolean hasExpressionAnnotation(AnnotatedTypeMirror atm) {
+        if (atm == null) return false;
```

Comment:
To be consistent in style with the similar code on line 671, also put this body in its own block.